vncviewer/touch.cxx

```diff
@@ -0,0 +1,627 @@
+  void fakeMotionEvent(const GHEvent origEvent);
+  void fakeButtonEvent(bool press, int button,
+                       const GHEvent origEvent);
+  void fakeKeyEvent(bool press, int keycode,
+                    const GHEvent origEvent);
+
+  virtual void handleGestureEvent(const GHEvent& event);
+
+private:
+  Window wnd;
+  int touchButtonMask;
+  int lastDistance;
+};
+
+static std::map<Window, class FLTKGestureHandler*> handlers;
+
+FLTKGestureHandler::FLTKGestureHandler(Window wnd)
+  : wnd(wnd), touchButtonMask(0)
+{
+}
+
+/*
+void FLTKGestureHandler::setTouchEventState(XEvent* dst)
+{
+  dst->xbutton.state |= touchButtonMask;
+}
+*/
+
+void FLTKGestureHandler::processEvent(const XIDeviceEvent* devev)
+{
+  switch (devev->evtype) {
+  case XI_Motion:
+    fakeMotionEvent(devev);
+    break;
+  case XI_ButtonPress:
+    fakeButtonEvent(true, devev->detail, devev);
+    break;
+  case XI_ButtonRelease:
+    fakeButtonEvent(false, devev->detail, devev);
+    break;
+  case XI_TouchBegin:
+  case XI_TouchUpdate:
+  case XI_TouchEnd:
+    registerEvent(devev);
+    break;
+  }
+}
+
+void FLTKGestureHandler::prepXEvent(XEvent* dst, const XIDeviceEvent* src)
+{
+  // XButtonEvent and XMotionEvent are almost identical, so we
+  // don't have to care which it is for these fields
+  dst->xbutton.serial = src->serial;
+  dst->xbutton.display = src->display;
+  dst->xbutton.window = src->event;
+  dst->xbutton.root = src->root;
+  dst->xbutton.subwindow = src->child;
+  dst->xbutton.time = src->time;
+  dst->xbutton.x = src->event_x;
+  dst->xbutton.y = src->event_y;
+  dst->xbutton.x_root = src->root_x;
+  dst->xbutton.y_root = src->root_y;
+  dst->xbutton.state = src->mods.effective;
+  dst->xbutton.state |= ((src->buttons.mask[0] >> 1) & 0x1f) << 8;
+  dst->xbutton.same_screen = True; // FIXME
+}
+
+void FLTKGestureHandler::prepXEvent(XEvent* dst, const GHEvent src)
+{
+  Window root, child;
+  int x_root, y_root;
+  XkbStateRec state;
+
+  // We don't have a real event to steal things from, so we'll have
+  // to fake these events based on the current state of things
+
+  root = XDefaultRootWindow(fl_display);
+  XTranslateCoordinates(fl_display, wnd, root,
+                        src.event_x,
+                        src.event_y,
+                        &x_root, &y_root, &child);
+  XkbGetState(fl_display, XkbUseCoreKbd, &state);
+
+  // XButtonEvent and XMotionEvent are almost identical, so we
+  // don't have to care which it is for these fields
+  dst->xbutton.serial = XLastKnownRequestProcessed(fl_display);
+  dst->xbutton.display = fl_display;
+  dst->xbutton.window = wnd;
+  dst->xbutton.root = root;
+  dst->xbutton.subwindow = None;
+  dst->xbutton.time = fl_event_time;
+  dst->xbutton.x = src.event_x;
+  dst->xbutton.y = src.event_y;
+  dst->xbutton.x_root = x_root;
+  dst->xbutton.y_root = y_root;
+  dst->xbutton.state = state.mods;
+  dst->xbutton.state |= ((state.ptr_buttons >> 1) & 0x1f) << 8; // FIXME: Check this!
```

Comment:
Did you check that fixme? :)

Reply:
I have no idea what that code is supposed to do :)

Reply:
The `state` field in the pointer events (and keyboard for that matter) is a combination of the keyboard state and pressed mouse buttons. However, what we get from XKB is separated in to multiple fields. So that line tries to combine them.

I just took a guess at it, so that calculation needs to be verified that it is indeed correct. :)

Reply:
I did some printf debugging of what the final value of `xbutton.state` was, and it seems to look okay. The outputted values match with the mask definitions in X.h, so I'm saying it works.

26217b26